content/advent-2015/etcd-distributed-key-value-store-with-grpc-http2.md

```diff
@@ -0,0 +1,362 @@
++++
+author = ["Gyu-Ho Lee"]
+date = "2015-12-29T08:00:00+00:00"
+linktitle = "etcd: distributed key-value store with grpc/http2"
+series = ["Advent 2015"]
+title = "etcd: distributed key-value store with grpc/http2"
++++
+
+## What is etcd?
+
+<img alt="etcd"
+     src="https://raw.githubusercontent.com/coreos/etcd/master/logos/etcd-horizontal-color.png"
+     style="float:right; margin-bottom:2em;"/>
+
+etcd is a distributed, consistent key-value store, written in Go. Linux OS uses `/etc` directory to store its system configurations. etcd provides more reliable ways to store such data. It is distributed by replicating data to multiple machines, therefore highly available against single point of failures. Using Raft consensus algorithms, etcd gracefully handles network partitions and machine failures, even leader failures. etcd is being widely used in production: CockroachDB, CoreOS, Kubernetes, etc.
+
+## How does etcd work?
+
+etcd cluster is leader-based. A leader is elected by other members in cluster. Once elected, the leader starts processing client requests and replicating them to its followers. All those server-to-server communications are done by RPC (Remote Procedure Call).
+
+## What matters to etcd?
+
+Latency matters. Data need to be delivered as fast as possible.  Memory/CPU usage matters. etcd needs to replicate messages with minimum resources. gRPC is one way that can reduce such costs and minimize latency.
```

Comment:
Maybe you could talk briefly about why these things are important - e.g. lower latency means higher throughput for the state machine, and consequently consistent operations, etc